src/cart/crt_hlc.c

```diff
@@ -0,0 +1,128 @@
+ *    this list of conditions, and the following disclaimer in the
+ *    documentation and/or materials provided with the distribution.
+ *
+ * 3. In addition, redistributions of modified forms of the source or binary
+ *    code must carry prominent notices stating that the original code was
+ *    changed and the date of the change.
+ *
+ * 4. All publications or advertising materials mentioning features or use of
+ *    this software are asked, but not required, to acknowledge that it was
+ *    developed by Intel Corporation and credit the contributors.
+ *
+ * 5. Neither the name of Intel Corporation, nor the name of any Contributor
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/**
+ * This file is part of CaRT. Hybrid Logical Clock (HLC) implementation.
+ */
+#include "crt_hlc.h"
+#include <gurt/atomic.h>
+#include <time.h>
+
+#if defined(HAVE_STDATOMIC)
+#if defined(__INTEL_COMPILER)
+#pragma warning(push)
+/**
+ * argument of type "_Atomic(uint64_t={__uint64_t={unsigned long}}) *" is
+ * incompatible with parameter of type "const volatile void *"
+ * (dropping qualifiers)
+ */
+#pragma warning(disable:2330)
+#endif /* __INTEL_COMPILER */
+#endif /* HAVE_STDATOMIC */
+
+#define CRT_HLC_MASK 0xFFFF
+
+static ATOMIC uint64_t crt_hlc;
+
+/** Get local physical time */
+static inline uint64_t crt_hlc_localtime_get(void)
+{
+	struct timespec now;
+	uint64_t	hlc;
+	int		rc;
+
+	rc = clock_gettime(CLOCK_REALTIME_COARSE, &now);
+	hlc = rc ? atomic_load_consume(&crt_hlc)
+		 : (now.tv_sec * 1e9 + now.tv_nsec);
+
+	return hlc & ~CRT_HLC_MASK;
+}
+
+int crt_hlc_send(uint64_t *time)
+{
+	uint64_t hlc, new;
+	uint64_t pt = crt_hlc_localtime_get();
+
+	do {
+		hlc = atomic_load_consume(&crt_hlc);
+		new = (hlc & ~CRT_HLC_MASK) < pt ? pt : (hlc + 1);
+	} while (!atomic_compare_exchange(&crt_hlc, hlc, new));
+
+	if (time != NULL)
+		*time = new;
+
+	return 0;
+}
+
+int crt_hlc_receive(uint64_t *time, uint64_t msg)
+{
+	uint64_t hlc, new, ml = msg & ~CRT_HLC_MASK;
+	uint64_t pt = crt_hlc_localtime_get();
+
+	do {
+		hlc = atomic_load_consume(&crt_hlc);
```

Comment:
Same comment above about atomics.

Reply:
My point is that I think it's a noop on intel but I could be wrong.